### PR TITLE
Add new evidence-related traits to Vitally rollup

### PR DESCRIPTION
**Pull request by cissyxyu**
## WHAT
Add new data points pertaining to Evidence Activities to Vitally data rollup for districts, schools, and teachers.

## WHY
Admin need this data to analyze Evidence activities.

## HOW
Roll up the data points with shared utility functions (Aggregated into a utility module VitallySharedFunctions) and some function performing specific Rails queries to fetch the data in question.

### Screenshots
(If applicable. Also, please censor any sensitive data)

### Notion Card Links
https://www.notion.so/quill/Vitally-Data-Update-Add-Reading-for-Evidence-Data-to-the-School-and-District-Level-53fbd58e4e914911a79ecc66142d9829?pvs=4

### What have you done to QA this feature?
I wrote specs to ensure the data is working properly. We unfortunately don't have a staging setup for Vitally so I will have to monitor Vitally stats for a few days to ensure the data is being properly passed to Vitally for admin.

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  yes
Have you deployed to Staging? | yes
Self-Review: Have you done an initial self-review of the code below on Github? | yes

### CODE REVIEW

File: services/QuillLMS/app/models/concerns/vitally_shared_functions.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  def sum_students(activities)
+    activities.map { |r| r&.assigned_student_ids&.count || 0 }.sum
+  end
+
+  def filter_evidence(activities)
+    evidence_ids = Activity.where(activity_classification_id: evidence_id).pluck(:id)
+    activities.select {|r| evidence_ids.include?(r.id) }
+  end
+
+  def evidence_id
```
Comment (brendanshean) [inline services/QuillLMS/app/models/concerns/vitally_shared_functions.rb]: I think this should be a constant.  `EVIDENCE_ID = ActivityClassification.evidence.id`

File: services/QuillLMS/app/services/vitally_integration/serialize_vitally_sales_account.rb [hunk 2]
```diff
@@ -64,6 +69,30 @@ def data
...
+      school_year_end = school_year_start + 1.year
+      active_students_this_year = active_students_query(@school).where("activity_sessions.completed_at >= ?", school_year_start).count
+
+      evidence_activities_assigned_all_time = evidence_assigned_count(@school)
+      evidence_activities_assigned_this_year = evidence_assigned_in_year_count(@school, school_year_start, school_year_end)
+      evidence_activities_completed_all_time = evidence_finished(@school).count
+      evidence_activities_completed_this_year = evidence_completed_in_year_count(@school, school_year_start, school_year_end)
+      evidence_activities_completed_per_student_this_year = activities_per_student(active_students_this_year, evidence_activities_completed_this_year)
+
+      {
```
Comment (brendanshean) [inline services/QuillLMS/app/services/vitally_integration/serialize_vitally_sales_account.rb:84]: You can leave out the values in the hash that have same variable name:

```
      {
        evidence_activities_assigned_all_time:,
        evidence_activities_assigned_this_year:,
        evidence_activities_assigned_last_year: get_from_cache('evidence_activities_assigned'),
        evidence_activities_completed_all_time:,
        evidence_activities_completed_this_year:,
        evidence_activities_completed_last_year: get_from_cache('evidence_activities_completed'),
        evidence_activities_completed_per_student_this_year:,
        evidence_activities_completed_per_student_last_year: get_from_cache("completed_evidence_activities_per_student"),
      }

File: services/QuillLMS/app/services/vitally_integration/serialize_vitally_sales_organization.rb [hunk 3]
```diff
@@ -76,6 +78,32 @@ def diagnostic_rollups
       }
     end
 
+    private def evidence_rollups
+      school_year_start = School.school_year_start(Time.current)
+      last_school_year_start = school_year_start - 1.year
+      school_year_end = school_year_start + 1.year
+      active_students_this_year = active_students(school_year_start)
+      active_students_last_year = active_students(last_school_year_start, school_year_start)
+
```
Comment (brendanshean) [inline services/QuillLMS/app/services/vitally_integration/serialize_vitally_sales_organization.rb:87]: For consistency, extract out all temp variables:
```
evidence_activities_assigned_all_time = evidence_assigned_count(@district)
evidence_activities_assigned_this_year = evidence_assigned_in_year_count(@district, school_year_start, school_year_end)
evidence_activities_assigned_last_year = evidence_assigned_in_year_count(@district, last_school_year_start, school_year_start)
evidence_activities_completed_all_time = evidence_finished(@district).count
evidence_activities_completed_this_year = evidence_completed_in_year_count(@district, school_year_start, school_year_end)
evidence_activities_completed_last_year = evidence_completed_in_year_count(@district, last_school_year_start, school_year_start)
evidence_activities_completed_per_student_this_year = activities_per_student(active_students_this_year, evidence_activities_completed_this_year)
evidence_activities_completed_per_student_last_year = activities_per_student(active_students_last_year, evidence_activities_completed_last_year)
```

Then you have a simpler resulting hash:
```
{
  evidence_activities_assigned_all_time:,
  evidence_activities_assigned_this_year:,
  evidence_activities_assigned_last_year:,
  evidence_activities_completed_all_time:,
  evidence_activities_completed_this_year:,
  evidence_activities_completed_last_year:,
  evidence_activities_completed_per_student_this_year:,
  evidence_activities_completed_per_student_last_year:
}
```

File: services/QuillLMS/app/services/vitally_integration/serialize_vitally_sales_user.rb [hunk 4]
```diff
@@ -92,7 +95,9 @@ def data
           diagnostics_finished_last_year: get_from_cache("diagnostics_finished"),
           percent_completed_diagnostics_this_year: diagnostics_assigned_this_year > 0 ? (diagnostics_finished_this_year.to_f / diagnostics_assigned_this_year).round(2) : 'N/A',
           percent_completed_diagnostics_last_year: get_from_cache("percent_completed_diagnostics"),
+          evidence_activities_assigned_all_time: evidence_activities_assigned_all_time,
```
Comment (brendanshean) [inline services/QuillLMS/app/services/vitally_integration/serialize_vitally_sales_user.rb]: Key is all that is needed: `evidence_activities_assigned_all_time:,`

File: services/QuillLMS/app/services/vitally_integration/serialize_vitally_sales_user.rb [hunk 5]
```diff
@@ -92,7 +95,9 @@ def data
...
           evidence_activities_assigned_this_year: evidence_activities_assigned_this_year,
+          evidence_activities_completed_all_time: evidence_activities_completed_all_time,
```
Comment (brendanshean) [inline services/QuillLMS/app/services/vitally_integration/serialize_vitally_sales_user.rb]: Key is all that is needed: `evidence_activities_completed_all_time:,`

**Review by brendanshean — APPROVED** [review]
I like the extraction into the concern.  Some minor comments but nothing requiring a re-review.

File: services/QuillLMS/app/models/concerns/vitally_school_stats.rb [hunk 6]
```diff
@@ -22,15 +22,20 @@ module VitallySchoolStats
...
-    end
+  def activities_assigned_query(school)
+    ClassroomUnit.joins("JOIN unit_activities ON classroom_units.unit_id=unit_activities.unit_id")
+    .joins("JOIN activities ON activities.id = unit_activities.activity_id")
+    .joins("JOIN classrooms ON classrooms.id = classroom_units.classroom_id")
+    .joins("JOIN classrooms_teachers ON classrooms.id=classrooms_teachers.classroom_id")
+    .joins("JOIN schools_users ON schools_users.user_id = classrooms_teachers.user_id")
+    .joins("JOIN schools ON schools_users.school_id=schools.id")
+    .where("schools.id = ?", school.id)
+    .select("assigned_student_ids, activities.id, unit_activities.created_at")
```
Comment (anathomical) [inline services/QuillLMS/app/models/concerns/vitally_school_stats.rb]: I don't know that this is better, but we can take advantage of all the `has_many` and other relationships that we've put into our model when making our join calls.  This should result in the same SQL, so you can just skip this, but I figured I'd call it out.:
```ruby
ClassroomUnit.joins(classroom: {teachers: :school}, unit: :activities)
  .where("schools.id = ?", school.id)
  .select("assigned_student_ids", "activities.id", "unit_activities.created_at")
```

File: services/QuillLMS/app/models/concerns/vitally_shared_functions.rb (hunk 1 above)
Comment (anathomical) [reply services/QuillLMS/app/models/concerns/vitally_shared_functions.rb]: Agreed.  By making it a constant, the database will be read once during start-up whenever the server parses the class definition, but then won't have to be read again for each use of the class.

File: services/QuillLMS/app/services/vitally_integration/serialize_vitally_sales_organization.rb [hunk 7]
```diff
@@ -165,6 +193,31 @@ def activities_completed_all_time
...
+    def activities_assigned_query(district)
+      ClassroomUnit.joins("JOIN unit_activities ON classroom_units.unit_id=unit_activities.unit_id")
+      .joins("JOIN activities ON activities.id = unit_activities.activity_id")
+      .joins("JOIN classrooms ON classrooms.id = classroom_units.classroom_id")
+      .joins("JOIN classrooms_teachers ON classrooms.id=classrooms_teachers.classroom_id")
+      .joins("JOIN schools_users ON schools_users.user_id = classrooms_teachers.user_id")
+      .joins("JOIN schools ON schools_users.school_id=schools.id")
+      .joins("JOIN districts ON schools.district_id = districts.id")
+      .where("districts.id = ?", district.id)
+      .select("assigned_student_ids, activities.id, unit_activities.created_at")
```
Comment (anathomical) [inline services/QuillLMS/app/services/vitally_integration/serialize_vitally_sales_organization.rb]: Same as a previous comment: this is probably re-writeable using defined model relationships.

File: services/QuillLMS/app/services/vitally_integration/serialize_vitally_sales_organization.rb [hunk 8]
```diff
@@ -165,6 +193,31 @@ def activities_completed_all_time
...
+      ClassroomsTeacher.joins("JOIN users ON users.id=classrooms_teachers.user_id")
+        .joins("JOIN schools_users ON schools_users.user_id=users.id")
+        .joins("JOIN schools ON schools.id=schools_users.school_id")
+        .joins("JOIN classrooms ON classrooms.id=classrooms_teachers.classroom_id")
+        .joins("JOIN classroom_units ON classroom_units.classroom_id=classrooms.id")
+        .joins("JOIN activity_sessions ON activity_sessions.classroom_unit_id=classroom_units.id")
+        .joins("JOIN activities ON activity_sessions.activity_id = activities.id")
+        .joins("JOIN districts ON schools.district_id = districts.id")
+        .where("districts.id = ?", district.id)
+        .where('activity_sessions.state = ?', 'finished')
```
Comment (anathomical) [inline services/QuillLMS/app/services/vitally_integration/serialize_vitally_sales_organization.rb]: Again with the possibly simpler-to-read JOIN syntax.

File: services/QuillLMS/spec/services/vitally_integration/serialize_vitally_sales_account_spec.rb [hunk 9]
```diff
@@ -22,11 +22,15 @@
   let(:subscription) { create(:subscription, account_type: Subscription::SCHOOL_PAID) }
 
   before do
+    create(:activity_classification, key: 'evidence')
```
Comment (anathomical) [inline services/QuillLMS/spec/services/vitally_integration/serialize_vitally_sales_account_spec.rb]: I'm pretty sure we have a convenience factory for this and that you can just call `create(:evidence)`

File: services/QuillLMS/spec/services/vitally_integration/serialize_vitally_sales_organization_spec.rb [hunk 10]
```diff
@@ -4,6 +4,10 @@
 
 describe VitallyIntegration::SerializeVitallySalesOrganization do
 
+  before do
+    create(:activity_classification, key: 'evidence')
```
Comment (anathomical) [inline services/QuillLMS/spec/services/vitally_integration/serialize_vitally_sales_organization_spec.rb]: Again, I'm pretty sure that `create(:evidence)` is defined in the ActivityClassification factory.

File: services/QuillLMS/app/models/concerns/vitally_shared_functions.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+module VitallySharedFunctions
+  extend ActiveSupport::Concern
+
+  def activities_per_student(active_students, activities_finished)
+    if active_students > 0
+      (activities_finished.to_f / active_students).round(2)
+    else
+      0
+    end
```
Comment (anathomical) [inline services/QuillLMS/app/models/concerns/vitally_shared_functions.rb]: I would do this with an early return:
```ruby
return 0 unless active_students.nonzero?

(activities.finished.to_f / active_students).round(2)
```

**Review by anathomical — APPROVED** [review]
Looks good.  Some style suggestions you can take or leave, but I'm impressed that we've managed to keep Vitally calculations as cleanly-organized as we have.

**Pull request merged**